From 33469fe1be59affd9b4a7a9562dd1ff4862323e1 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Thu, 1 Dec 2016 20:56:25 +0000
Subject: [PATCH] [CONFIGURATION-641] Improved exception when storing
 XMLPropertyListConfiguration.

It is now checked whether the locator has been correctly initialized.
If not, an exception with a helpful message is thrown.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1772268 13f79535-47bb-0310-9956-ffa450edef68
---
 .../plist/XMLPropertyListConfiguration.java   |  6 +++
 .../TestXMLPropertyListConfiguration.java     | 37 +++++++++++++++++--
 2 files changed, 40 insertions(+), 3 deletions(-)

diff --git a/src/main/java/org/apache/commons/configuration2/plist/XMLPropertyListConfiguration.java b/src/main/java/org/apache/commons/configuration2/plist/XMLPropertyListConfiguration.java
index 07da3ce4f0..7895a5de26 100644
--- a/src/main/java/org/apache/commons/configuration2/plist/XMLPropertyListConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/plist/XMLPropertyListConfiguration.java
@@ -261,6 +261,12 @@ public InputSource resolveEntity(String publicId, String systemId)
     @Override
     public void write(Writer out) throws ConfigurationException
     {
+        if (locator == null)
+        {
+            throw new ConfigurationException("Save operation not properly "
+                    + "initialized! Do not call write(Writer) directly,"
+                    + " but use a FileHandler to save a configuration.");
+        }
         PrintWriter writer = new PrintWriter(out);
 
         if (locator.getEncoding() != null)
diff --git a/src/test/java/org/apache/commons/configuration2/plist/TestXMLPropertyListConfiguration.java b/src/test/java/org/apache/commons/configuration2/plist/TestXMLPropertyListConfiguration.java
index 3b35a663bf..72afdeb034 100644
--- a/src/test/java/org/apache/commons/configuration2/plist/TestXMLPropertyListConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/plist/TestXMLPropertyListConfiguration.java
@@ -23,18 +23,19 @@
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
 
 import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
 import java.io.StringWriter;
+import java.io.Writer;
 import java.util.Arrays;
 import java.util.Calendar;
 import java.util.Iterator;
 import java.util.List;
 import java.util.TimeZone;
 
-import junitx.framework.ArrayAssert;
-import junitx.framework.ListAssert;
-import junitx.framework.ObjectAssert;
 import org.apache.commons.configuration2.Configuration;
 import org.apache.commons.configuration2.ConfigurationAssert;
 import org.apache.commons.configuration2.ConfigurationComparator;
@@ -46,6 +47,10 @@
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
 
+import junitx.framework.ArrayAssert;
+import junitx.framework.ListAssert;
+import junitx.framework.ObjectAssert;
+
 /**
  * @author Emmanuel Bourg
  * @version $Id$
@@ -508,4 +513,30 @@ public void testSaveArray() throws ConfigurationException
 
         checkArrayProperty(Arrays.asList(elems));
     }
+
+    /**
+     * Tests a direct invocation of the write() method. This test is
+     * related to CONFIGURATION-641.
+     */
+    @Test
+    public void testWriteCalledDirectly() throws IOException
+    {
+        config = new XMLPropertyListConfiguration();
+        config.addProperty("foo", "bar");
+
+        Writer out = new FileWriter(folder.newFile());
+        try
+        {
+            config.write(out);
+            fail("No exception thrown!");
+        }
+        catch (ConfigurationException e)
+        {
+            assertThat(e.getMessage(), containsString("FileHandler"));
+        }
+        finally
+        {
+            out.close();
+        }
+    }
 }
